pkg/ingester/ingester_v2.go

```diff
@@ -334,3 +352,34 @@ func (i *Ingester) getOrCreateTSDB(userID string) (*tsdb.DB, error) {
 }
 
 func (i *Ingester) userDir(userID string) string { return filepath.Join(i.cfg.TSDBConfig.Dir, userID) }
+
+func (i *Ingester) closeAllTSDB() {
+	i.userStatesMtx.Lock()
+
+	wg := &sync.WaitGroup{}
+	wg.Add(len(i.TSDBState.dbs))
+
+	// Concurrently close all users TSDB
+	for userID, db := range i.TSDBState.dbs {
+		userID := userID
+
+		go func(db *tsdb.DB) {
+			defer wg.Done()
+
+			if err := db.Close(); err != nil {
+				level.Warn(util.Logger).Log("msg", "unable to close TSDB", "err", err, "user", userID)
+				return
+			}
+
+			// Now that the TSDB has been closed, we should remove it from the
+			// set of open ones.
+			i.userStatesMtx.Lock()
```

Comment:
We already acquire the lock at the beginning of the parent function this is going to deadlock no?

EDIT: No, they'll just block until the loop exits, but maybe add a comment?

Reply:
As you stated, there's no deadlock. They block, but realistically the `for` loop is fast and the lock is acquired after `db.Close()` which may be a slower operation.

Reply:
> but maybe add a comment?

Done